nltk/tokenize/sonority_sequencing.py

```diff
@@ -0,0 +1,141 @@
+# Natural Language Toolkit: Tokenizers
+#
+# Copyright (C) 2001-2019 NLTK Project
+# Author: Christopher Hench <chris.l.hench@gmail.com>
+#         Alex Estes
+# URL: <http://nltk.sourceforge.net>
+# For license information, see LICENSE.TXT
+
+"""
+The Sonority Sequencing Principle (SSP) is a language agnostic algorithm proposed
+by Otto Jesperson in 1904. The sonorous quality of a phoneme is judged by the
+openness of the lips. Syllable breaks occur before troughs in sonority. For more
+on the SSP see Selkirk (1984).
+
+The default implementation uses the English alphabet, but the `sonority_hiearchy`
+can be modified to IPA or any other alphabet for the use-case. The SSP is a
+universal syllabification algorithm, but that does not mean it performs equally
+across languages. Bartlett et al. (2009) is a good benchmark for English accuracy
+if utilizing IPA (pg. 311).
+
+Importantly, if a custom hiearchy is supplied and vowels span across more than
+one level, they should be given separately to the `vowels` class attribute.
+
+References:
+- Otto Jespersen. 1904. Lehrbuch der Phonetik.
+  Leipzig, Teubner. Chapter 13, Silbe, pp. 185-203.
+- Elisabeth Selkirk. 1984. On the major class features and syllable theory.
+  In Aronoff & Oehrle (eds.) Language Sound Structure: Studies in Phonology.
+  Cambridge, MIT Press. pp. 107-136.
+- Susan Bartlett, et al. 2009. On the Syllabification of Phonemes.
+  In HLT-NAACL. pp. 308-316.
+"""
+
+from __future__ import unicode_literals
+
+import re
+
+from nltk.tokenize.api import TokenizerI
+from nltk.util import ngrams
+
+
+class SyllableTokenizer(TokenizerI):
+    """
+    Syllabifies words based on the Sonority Sequencing Principle (SSP).
+
+    >>> SSP = SyllableTokenizer()
+    >>> SSP.tokenize('justification')
+    ['jus', 'ti', 'fi', 'ca', 'tion']
+    """
+
+    def __init__(self, lang='en', sonority_hierarchy=False):
+        # Sonority hierarchy should be provided in descending order.
+        # If vowels are spread across multiple levels, they should be
+        # passed assigned self.vowels var together, otherwise should be
+        # placed in first index of hierarchy.
+        if not sonority_hierarchy and lang == 'en':
+            sonority_hierarchy = ['aeiouy',  # vowels.
+                                  'lmnrw',  # nasals.
+                                  'zvsf',  # fricatives.
+                                  'bcdgtkpqxhj'  # stops.
+                                  ]
+
+        self.vowels = sonority_hierarchy[0]
+        self.phoneme_map = {}
+        for i, level in enumerate(sonority_hierarchy):
+            for c in level:
+                self.phoneme_map[c] = len(sonority_hierarchy) - i
+
```

Comment:
add
```suggestion
self.phoneme_map[c.upper()] = len(sonority_hierarchy) - i
```

Reply:
👍 this should be done to handle the `ß` issue. Would it be better to:

```python
self.phoneme_map[c.upper().lower()] = len(sonority_hierarchy) - i
```

Reply:
Hmmm but actually that complicates the access, you can loop through character by character after `ß -> ss` =)